### ISSUE TITLE
warnings in core/time.h (PR #4926) -- is_leap_year / days_in_month declared static' but never defined

### ISSUE THREAD

**Issue by bkaufx** [esphome/issues#4581]
### The problem

Warnings being received after PR 4926: https://github.com/esphome/esphome/pull/4926

Issue exists for ESP8266 and ESP32 with ESP-IDF framework, but not ESP32 with Arduino framework.

I tried to play around and figure out how to fix this but no luck.  Somehow core/time.cpp isn't being included properly?

### Which version of ESPHome has the issue?

2023.6.0

### What type of installation are you using?

pip

### Which version of Home Assistant has the issue?

n/a

### What platform are you using?

ESP8266

### Component causing the issue

core/time.h

### Example YAML snippet

```yaml
esphome:
  name: asdf

esp8266:
  board: esp01_1m
```

### Anything in the logs that might be useful for us?

```txt
Compiling .pioenvs/asdf/FrameworkArduino/cont_util.cpp.o
In file included from src/esphome.h:26,
                 from src/main.cpp:3:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
   11 | static bool is_leap_year(uint32_t year);
      |             ^~~~~~~~~~~~
Compiling .pioenvs/asdf/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
   13 | static uint8_t days_in_month(uint8_t month, uint16_t year);
      |                ^~~~~~~~~~~~~
```

**Comment by ssieb** [comment]
This has nothing to do with the sensor.  It's something specific to ESP8266 in the `time.h` file.  It doesn't happen with ESP32.

**Comment by droscy** [comment]
@ssieb it happens also on ESP32 with esp-idf framework.

```
Compiling .pioenvs/esp07/src/esphome/components/wifi/wifi_component_esp_idf.o
In file included from src/esphome/components/time/real_time_clock.h:8,
                 from src/esphome/components/time/real_time_clock.cpp:1:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
 static bool is_leap_year(uint32_t year);
             ^~~~~~~~~~~~
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
 static uint8_t days_in_month(uint8_t month, uint16_t year);
                ^~~~~~~~~~~~~
```

**Comment by ssieb** [comment]
I wonder why it doesn't happen with an arduino compile.

**Comment by bkaufx** [comment]
OK, good title change.  The problem is definitely a general one and not specific to total_daily_energy.  Even just using this as the entire config still gives the same warnings.

YAML config:
```
esphome:
  name: asdf

esp8266:
  board: esp01_1m
```

Warnings:
```
Compiling .pioenvs/asdf/FrameworkArduino/cont_util.cpp.o
In file included from src/esphome.h:26,
                 from src/main.cpp:3:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
   11 | static bool is_leap_year(uint32_t year);
      |             ^~~~~~~~~~~~
Compiling .pioenvs/asdf/FrameworkArduino/core_esp8266_app_entry_noextra4k.cpp.o
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
   13 | static uint8_t days_in_month(uint8_t month, uint16_t year);
      |                ^~~~~~~~~~~~~
```

**Comment by tommyguuuun** [comment]
I got the same warnings since the newest update

**Comment by stackner** [comment]
Same happening here

**Comment by RobvanHamersveld** [comment]
Same here, I looks like everything works fine though.

**Comment by mspiridakis** [comment]
Same here, it must be a bug. I don't see any changes in docs for time component.

**Comment by KitHubek** [comment]
after update from 2023.6 to 2023.6.1 problem still exsist, at log:

```
Compiling /data/wt32-eth01-bt-proxy/.pioenvs/wt32-eth01-bt-proxy/bootloader_support/src/flash_qio_mode.o
In file included from src/esphome.h:69,
                 from src/main.cpp:3:
src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
 static bool is_leap_year(uint32_t year);
             ^~~~~~~~~~~~
src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
 static uint8_t days_in_month(uint8_t month, uint16_t year);
                ^~~~~~~~~~~~~
Compiling /data/wt32-eth01-bt-proxy/.pioenvs/wt32-eth01-bt-proxy/bootloader_support/src/bootloader_flash_config_esp32.o
```

**Comment by mspiridakis** [comment]
> after update from 2023.6 to 2023.6.1 problem still exsist, at log:
> 
> ```
> Compiling /data/wt32-eth01-bt-proxy/.pioenvs/wt32-eth01-bt-proxy/bootloader_support/src/flash_qio_mode.o
> In file included from src/esphome.h:69,
>                  from src/main.cpp:3:
> src/esphome/core/time.h:11:13: warning: 'bool esphome::is_leap_year(uint32_t)' declared 'static' but never defined [-Wunused-function]
>  static bool is_leap_year(uint32_t year);
>              ^~~~~~~~~~~~
> src/esphome/core/time.h:13:16: warning: 'uint8_t esphome::days_in_month(uint8_t, uint16_t)' declared 'static' but never defined [-Wunused-function]
>  static uint8_t days_in_month(uint8_t month, uint16_t year);
>                 ^~~~~~~~~~~~~
> Compiling /data/wt32-eth01-bt-proxy/.pioenvs/wt32-eth01-bt-proxy/bootloader_support/src/bootloader_flash_config_esp32.o
> ```

issue will be resolved in 2023.6.2 according to https://github.com/esphome/esphome/pull/4993

**Comment by KitHubek** [comment]
yes, ok. thanks

**Comment by stackner** [comment]
It has been fixed in 2023.6.2 no more errors.
